trie/sync.py

```diff
@@ -0,0 +1,179 @@
+import bisect
+
+from eth_utils import encode_hex
+
+from trie.constants import (
+    NODE_TYPE_BRANCH,
+    NODE_TYPE_EXTENSION,
+    NODE_TYPE_LEAF,
+)
+from trie.exceptions import (
+    SyncRequestAlreadyProcessed,
+)
+from trie.utils.nodes import (
+    decode_node,
+    get_node_type,
+    is_blank_node,
+)
+from trie.utils.sha3 import keccak
+
+
+class SyncRequest:
+
+    def __init__(self, node_key, parent, depth, leaf_callback, is_raw=False):
+        """Create a new SyncRequest for a given HexaryTrie node.
+
+        :param node_key: The node's key.
+        :param parent: The node's parent.
+        :param depth: The ndoe's depth in the trie.
+        :param leaf_callback: A callback called when for all leaf children of this node.
+        :param is_raw: If True, HexaryTrieSync will simply store the node's data in the db,
+        without decoding and scheduling requests for children. This is needed to fetch contract
+        code when doing a state sync.
+        """
+        self.node_key = node_key
+        self.parents = []
+        if parent is not None:
+            self.parents = [parent]
+        self.depth = depth
+        self.leaf_callback = leaf_callback
+        self.is_raw = is_raw
+        self.dependencies = 0
+        self.data = None
+
+    def __lt__(self, other):
+        return self.depth < other.depth
+
+    def __repr__(self):
+        return "SyncRequest(%s, depth=%d)" % (encode_hex(self.node_key), self.depth)
+
+
+class HexaryTrieSync:
+
+    def __init__(self, root_hash, db, logger):
+        self.queue = []
+        self.requests = {}
+        self.db = db
+        self.root_hash = root_hash
+        self.logger = logger
+        self.schedule(root_hash, parent=None, depth=0, leaf_callback=self.leaf_callback)
+
+    def leaf_callback(self, data, parent):
+        """Called when we reach a leaf node.
+
+        Should be implemented by subclasses that need to perform special handling of leaves.
+        """
+
+    def next_batch(self, n=1):
+        """Return the next requests that should be dispatched."""
+        if len(self.queue) == 0:
+            return []
+        batch = self.queue[-n:]
```

Comment:
Maybe perform the `list(reversed())` here instead of in the return statement?  I'm not sure if it actually achieves any performance gains, but it seems like it avoids double allocating this value.